api/types/plugin/capability.go

```diff
@@ -0,0 +1,55 @@
+package plugin
+
+import (
+	"bytes"
+	"encoding"
+	"fmt"
+	"strings"
+)
+
+type CapabilityID struct {
+	Capability string
+	Prefix     string
+	Version    string
+}
+
+var (
+	_ fmt.Stringer             = CapabilityID{}
+	_ encoding.TextUnmarshaler = (*CapabilityID)(nil)
+	_ encoding.TextMarshaler   = CapabilityID{}
+)
+
+// String implements [fmt.Stringer] for CapabilityID
+func (t CapabilityID) String() string {
+	return fmt.Sprintf("%s.%s/%s", t.Prefix, t.Capability, t.Version)
+}
+
+// UnmarshalText implements [encoding.TextUnmarshaler] for CapabilityID
+func (t *CapabilityID) UnmarshalText(p []byte) error {
+	fqcap, version, _ := bytes.Cut(p, []byte{'/'})
+	idx := bytes.LastIndexByte(fqcap, '.')
+	if idx < 0 {
+		t.Prefix = ""
+		t.Capability = string(fqcap)
+	} else {
+		t.Prefix = string(fqcap[:idx])
+		t.Capability = string(fqcap[idx+1:])
+	}
+	t.Version = string(version)
+	return nil
+}
+
+// MarshalText implements [encoding.TextMarshaler] for CapabilityID
+func (t CapabilityID) MarshalText() ([]byte, error) {
+	// Assert that the value can be round-tripped successfully.
+	if strings.Contains(t.Capability, ".") {
+		return nil, fmt.Errorf("capability %q cannot contain a dot", t.Capability)
+	}
```

Comment:
Interesting; getting a failure in some e2e test in the CLI - not unlikely that the dummy value it used was _supposed_ to be valid 🤔 ;

```
=== Failed
=== FAIL: e2e/global TestPromptExitCode/plugin_install (0.28s)
    cli_test.go:203: assertion failed: error is not nil: json: error calling MarshalText for type plugin.CapabilityID: capability "docker.dummy/1.0" cannot contain a dot

=== FAIL: e2e/global TestPromptExitCode/plugin_upgrade (0.26s)
    cli_test.go:203: assertion failed: error is not nil: json: error calling MarshalText for type plugin.CapabilityID: capability "docker.dummy/1.0" cannot contain a dot
```

Reply:
Yes; `PluginInterfaceType` was frequently misused by setting the `Capability` field to the fully-qualified capability ID and leaving the other two fields blank. A value such as `PluginInterfaceType{Capability: "docker.dummy/1.0"}` would marshal to the string `".docker.dummy/1.0/"` which is not well-formed and [does not round-trip to the same value even with the old parser](https://go.dev/play/p/KFSl2GVK68D)